### ISSUE TITLE
解决训练过程验证map为0问题

### ISSUE THREAD

**Issue by Mr-zhao765** [bubbliiiing/yolov8-pytorch#24]
顺着B导的代码过一遍流程啥也没看出来，为什么从头训练评估的map值一直为0，最后按照yolov8的做法，DFL部分单独放到解码部分，还真解决了这个问题，怀疑是解码有问题，但看代码两个是等价的，希望B导在看看指导一下啊。
以下是在utils/callbacks.py 添加的代码替换outputs = self.bbox_util.decode_box(outputs)这个解码部分，同时预测和计算map的yolo.py文件也要一同修改。
        self.reg_max = 16
        self.no = self.num_classes + self.reg_max * 4
        self.stride = torch.Tensor([8, 16, 32])
        self.dfl = DFL(self.reg_max).to(torch.device("cuda"))
    def decode_boxes(self, outputs):
        dbox, cls, origin_cls, anchors, strides = outputs
        outputs = origin_cls
        shape = outputs[0].shape
        # anchor[2, 21504]  strides[1, 21504]
        anchors, strides = (x.transpose(0, 1) for x in make_anchors(outputs, self.stride, 0.5))
        # [1, 65, 21504]
        x_cat = torch.cat([xi.view(shape[0], self.no, -1) for xi in outputs], 2)
        box, cls = x_cat.split((self.reg_max * 4, self.num_classes), 1)  # [1, 64, 21504] [1, 1, 21504]
        # xywh
        dbox = dist2bbox(self.dfl(box), anchors.unsqueeze(0), xywh=True, dim=1) * strides
        dbox = dbox.permute(0, 2, 1) / self.input_shape[0]  # 归一化
        cls = cls.sigmoid().permute(0, 2, 1)
        y = torch.cat((dbox, cls), 2)
        return y

**Comment by bubbliiiing** [comment]
不应该呀。

**Comment by bubbliiiing** [comment]
你可以给我具体的代码吗？因为我在COCO基础上训练VOC是有的

**Comment by bubbliiiing** [comment]
感觉很神奇那

**Comment by Mr-zhao765** [comment]
在up上代码改了一下，感觉Yolov8的recall值有点拉啊，整体map还是很强的
链接：https://pan.baidu.com/s/1kziIxp0KW9uH8lHvxjbMzw 
提取码：1234

**Comment by bubbliiiing** [comment]
我周末看看

**Comment by bubbliiiing** [comment]
我试了好多次感觉好像没有什么差别呀，有没有数据可以共享呀0 0

**Comment by bubbliiiing** [comment]
如果方便的话，我试了一下，怎么都没看出差别
我看了看DFL里面的，权值也没有变化

**Comment by Milkyway-xX** [comment]
我也遇到了这个问题，把框框画出来发现左上右下坐标一样
类似这样。。。我只是换了个Backbone，其他的都没改
b'person 0.90' 287 189 287 188
b'person 0.89' 262 214 262 213
b'person 0.89' 262 164 262 163
b'person 0.88' 262 189 262 188
b'person 0.87' 287 214 287 213
b'person 0.84' 237 164 237 163
b'person 0.83' 237 214 237 213
b'person 0.82' 237 189 237 188
b'person 0.78' 262 239 262 239

**Comment by Milkyway-xX** [comment]
这个问题好像在原生的Yolov8上不存在，我试了一下，但是动了网络结构（我换了Backbone）过后好像就有了。

**Comment by Mr-zhao765** [comment]
可能bug吧，我这里这样改从头训练是有验证结果的

**Comment by wkgit88** [comment]
> 

您好，我也遇到这个问题了，把backbone换成了resnet，训练时训练集和验证集loss下降都没什么问题，但测试map是发现map都是0，一看发现所有的输出都成一个点了（左上右下坐标几乎一样）。请问您后来解决了吗

**Comment by Milkyway-xX** [comment]
> > 
> 
> 您好，我也遇到这个问题了，把backbone换成了resnet，训练时训练集和验证集loss下降都没什么问题，但测试map是发现map都是0，一看发现所有的输出都成一个点了（左上右下坐标几乎一样）。请问您后来解决了吗

按照楼主的方法解决就行

**Comment by wkgit88** [comment]
> 

谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥

**Comment by bubbliiiing** [comment]
我查了一下问题，发现是pretrained=False时，随机初始化会导致DFL随机初始化为小数
已经修正

**Comment by shuoshuoshuoer** [comment]
> > 
> 
> 谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥

您好打扰了，我的也是动了backbone，map一直为0，请问你是怎么解决的呢，我用楼主的也是有问题

**Comment by wkgit88** [comment]
> 

现在可以直接按up给的nets/yolo.py和trian.py中修改的来就行了，本质是dfl的初始化问题，你把前面说的那两个文件替换一下就行（对着改也可以，就两行 找 weight_init()

**Comment by wkgit88** [comment]
> > 谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥
> 
> 您好打扰了，我的也是动了backbone，map一直为0，请问你是怎么解决的呢，我用楼主的也是有问题

现在可以直接按up给的nets/yolo.py和trian.py中修改的来就行了，本质是dfl的初始化问题，你把前面说的那两个文件替换一下就行（对着改也可以，就两行 找 weight_init()

**Comment by Mengb001** [comment]
> 顺着B导的代码过一遍流程啥也没看出来，为什么从头训练评估的map值一直为0，最后按照yolov8的做法，DFL部分单独放到解码部分，还真解决了这个问题，怀疑是解码有问题，但看代码两个是等价的，希望B导在看看指导一下啊。 以下是在utils/callbacks.py 添加的代码替换outputs = self.bbox_util.decode_box(outputs)这个解码部分，同时预测和计算map的yolo.py文件也要一同修改。 self.reg_max = 16 self.no = self.num_classes + self.reg_max * 4 self.stride = torch.Tensor([8, 16, 32]) self.dfl = DFL(self.reg_max).to(torch.device("cuda")) def decode_boxes(self, outputs): dbox, cls, origin_cls, anchors, strides = outputs outputs = origin_cls shape = outputs[0].shape # anchor[2, 21504] strides[1, 21504] anchors, strides = (x.transpose(0, 1) for x in make_anchors(outputs, self.stride, 0.5)) # [1, 65, 21504] x_cat = torch.cat([xi.view(shape[0], self.no, -1) for xi in outputs], 2) box, cls = x_cat.split((self.reg_max * 4, self.num_classes), 1) # [1, 64, 21504] [1, 1, 21504] # xywh dbox = dist2bbox(self.dfl(box), anchors.unsqueeze(0), xywh=True, dim=1) * strides dbox = dbox.permute(0, 2, 1) / self.input_shape[0] # 归一化 cls = cls.sigmoid().permute(0, 2, 1) y = torch.cat((dbox, cls), 2) return y

你好。我在b导的代码上更换了主干且导入新主干的预训练权重，训练过程中loss下降到3，ap一直为零。即便使用更新修复过的代码也依然如此。按照你的建议，我修改了decode boxs代码，发现很快就有了期望的ap，但事情显然不是那么简单。我有10个类别，但是只有1个类别ap正常，其他全是0。loss在45左右。 打印结构，发现只有第一个类别正常，其他类别依然是一个点。一脸懵逼。

**Comment by Mengb001** [comment]
> 



> > 
> 
> 谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥

确实 按照楼主的方法直接修改decode bbox  原yolov8-s只有10的map值。我更换主干后，1万张数据集  评估ap依然是0  不知道为啥。

**Comment by bubbliiiing** [comment]
> > 
> 
> > > 
> > 
> > 
> > 谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥
> 
> 确实 按照楼主的方法直接修改decode bbox 原yolov8-s只有10的map值。我更换主干后，1万张数据集 评估ap依然是0 不知道为啥。
本质上我和他的过程是一样的呀0 0你还修改了什么吗

**Comment by Mengb001** [comment]
> > > 
> > 
> > 
> > > > 
> > > 
> > > 
> > > 谢谢，已经解决了，resnet可以使用了。但我试了一下按楼主的方法，原本yolov8-s的map下降的厉害，不知为啥
> > 
> > 
> > 确实 按照楼主的方法直接修改decode bbox 原yolov8-s只有10的map值。我更换主干后，1万张数据集 评估ap依然是0 不知道为啥。
> > 本质上我和他的过程是一样的呀0 0你还修改了什么吗

我只是更换了主干网络，换为swin transformer-tiny主干，然后自己加个adamW优化器，除此之外啥也没动。发现测试map一直是0  loss正常下降到3.其他设置都没动。    楼主的方法，会导致我第一个分类情况正常，其余类别map也一直是0，loss会变得很大，45左右。

**Comment by bubbliiiing** [comment]
可能是swin的问题0 0，如果不用swin可以做做测试，swin不好收敛

**Comment by successfulbarrier** [comment]
bicycle 0.0036 140 88 140 88
car 0.8846 262 154 262 154
car 0.8151 262 178 263 178
我也遇到了这个问题，感谢，感谢，V1版本的代码还是存在问题，建议UP更新
